File: lesson_15_20241126/frog_slow_fast.js
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

function detectLoopFloyd(head) {
    let fast = head;
    let slow = head;

    while(fast !==null && fast.next!==null) {
        slow = slow.next;
        fast = fast.next.next;
        if(slow === fast) {
            return true;
        }
    }
    return false;
}

function detectLoopFloydSteps(head, step) {
    let fast = head;
    let slow = head;

    while(fast !==null && fast.next!==null) {
        slow = slow.next;
        for(let i = 0; i<step; i++) { //перемещение быстрой лягушки
            fast = fast.next;
        }
        if(slow === fast) {
            return true;
        }
    }
    return false;
}

let head = new Node(1);
let second = new Node(2);
let third = new Node(3);
let fourth = new Node(4);

head.next = second;
second.next = third;
third.next = fourth;
fourth.next = second; // закольцуем

console.log("Закольцован? = "+detectLoopFloyd(head));
console.log("Закольцован? = "+detectLoopFloydSteps(head,3));

File: lesson_15_20241126/FrogSlowFast.java
package lesson_15_20241126;

public class FrogSlowFast {

    public static void main(String[] args) {
        Node head = new Node(1);
        Node second = new Node(2);
        Node third = new Node(3);
        Node fourth = new Node(4);

        head.next = second;
        second.next = third;
        third.next = fourth;
        fourth.next = second; // закольцуем

        System.out.println("Закольцован detectLoopFloyd ? = "+detectLoopFloyd(head));
        System.out.println("Закольцован detectLoopFloydSteps? = "+detectLoopFloydSteps(head,3));
    }


    static boolean detectLoopFloyd(Node head) {
        Node fast = head;
        Node slow = head;
    
        while(fast !=null && fast.next!=null) {
            slow = slow.next;
            fast = fast.next.next;
            if(slow == fast) {
                return true;
            }
        }
        return false;
    }

    static boolean detectLoopFloydSteps(Node head, int step) {
        Node fast = head;
        Node slow = head;
    
        while(fast !=null && fast.next!=null) {
            slow = slow.next;
            for(int i = 0; i<step; i++) { //перемещение быстрой лягушки
                fast = fast.next;
            }
            if(slow == fast) {
                return true;
            }
        }
        return false;
    }
}

class Node {
    int value;
    Node next;
    public Node(int value) {
        this.value = value;
    }
}
